### PR TITLE
Add Windows date picker

### PR DESCRIPTION
**Pull request by rectified95**
# Summary

Add Windows date picker. Closes microsoft/react-native-windows#4156
This PR extracts the DatePicker from React Native Windows in line with the lean core effort. It preserves the existing functionality available on Windows, rewriting the code using C++/WinRT to follow the latest recommended development practices.
Time picker will be added in a separate PR (tracked by microsoft/react-native-windows#4695)
### What's required for testing (prerequisites)?
Run the example app:
0. Install system requirements: https://microsoft.github.io/react-native-windows/docs/next/rnw-dependencies
1. Run`'npm run start:windows`
2. open example/windows/datePickerTest.sln in Visual Studio and run.

Or create your own test app and link this module:
https://microsoft.github.io/react-native-windows/docs/next/native-modules-setup#option-1-create-a-new-test-app

image

## Compatibility

| OS      | Implemented |
| ------- | :---------: |
| Windows |    ✅    |

## Checklist

- [X] I have tested this on a device and a simulator
- [x] I added the documentation in `README.md` 
- [ ] I mentioned this change in `CHANGELOG.md` (**doing**)
- [X] I updated the typed files (TS and Flow)
- [X] I added a sample use of the API in the example project (`example/App.js`)

### CODE REVIEW

File: windows/DateTimePickerWindows/DateTimePickerView.cpp [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+#include "pch.h"
+#include "JSValueXaml.h"
+#include "DateTimePickerView.h"
+#include "DateTimePickerView.g.cpp"
+
+namespace winrt {
+    using namespace Microsoft::ReactNative;
+    using namespace Windows::Foundation;
+    using namespace Windows::UI;
+    using namespace Windows::UI::Xaml;
```
Comment (asklar) [inline windows/DateTimePickerWindows/DateTimePickerView.cpp:13]: please uses aliases - ideally don't use using namespace on `Controls`, just reference `xaml::Controls::xyz` instead

File: windows/DateTimePickerWindows/DateTimePickerView.cpp [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+                {
+                    WriteProperty(eventDataWriter, L"newDate", timeInMilliseconds);
+                }
+                eventDataWriter.WriteObjectEnd();
+            });
+        }
+    }
+
+    winrt::DateTime DateTimePickerView::DateTimeFrom(int64_t timeInMilliSeconds, int64_t timeZoneOffsetInSeconds) {
+        auto timeInSeconds = timeInMilliSeconds / 1000;
```
Comment (asklar) [inline windows/DateTimePickerWindows/DateTimePickerView.cpp:146]: use const whenever possible

File: windows/DateTimePickerWindows/DateTimePickerView.cpp [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        time_t ttWithTimeZoneOffset = static_cast<time_t>(timeInSeconds) + timeZoneOffsetInSeconds;
+        winrt::DateTime dateTime = winrt::clock::from_time_t(ttWithTimeZoneOffset);
+
+        return dateTime;
+    }
+
+    int64_t DateTimePickerView::DateTimeToMiliseconds(winrt::DateTime dateTime, int64_t timeZoneOffsetInSeconds) {
+        time_t ttInSeconds = winrt::clock::to_time_t(dateTime);
+        auto timeInUtc = ttInSeconds - timeZoneOffsetInSeconds;
+        auto ttInMilliseconds = static_cast<int64_t>(timeInUtc) * 1000;
```
Comment (asklar) [inline windows/DateTimePickerWindows/DateTimePickerView.cpp:156]: check for overflow?

File: windows/DateTimePickerWindows/DateTimePickerView.h [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        DateTimePickerView(Microsoft::ReactNative::IReactContext const& reactContext);
+        void UpdateProperties(Microsoft::ReactNative::IJSValueReader const& reader);
+
+    private:
+        Microsoft::ReactNative::IReactContext m_reactContext{ nullptr };
+        bool m_updating{ false };
+        winrt::Windows::UI::Xaml::Controls::CalendarDatePicker::DateChanged_revoker m_dataPickerDateChangedRevoker{};
+
+        void RegisterEvents();
+        void OnDateChanged(winrt::Windows::Foundation::IInspectable const& sender, winrt::Windows::UI::Xaml::Controls::CalendarDatePickerDateChangedEventArgs const& args);
```
Comment (asklar) [inline windows/DateTimePickerWindows/DateTimePickerView.h:23]: alias it

File: windows/DateTimePickerWindows/DateTimePickerViewManager.cpp [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        nativeProps.Insert(L"timeZoneOffsetInSeconds", ViewManagerPropertyType::Number);
+
+        return nativeProps.GetView();
+    }
+
+    void DateTimePickerViewManager::UpdateProperties(FrameworkElement const& view,
+        IJSValueReader const& propertyMapReader) noexcept {
+        if (auto dateTimePickerView = view.try_as<DateTimePickerView>()) {
+            dateTimePickerView->UpdateProperties(propertyMapReader);
+        }
```
Comment (asklar) [inline windows/DateTimePickerWindows/DateTimePickerViewManager.cpp:57]: else? debug spew/assert

File: windows/DateTimePickerWindows/DateTimePickerWindows.vcxproj [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+      <Platform>Win32</Platform>
+    </ProjectConfiguration>
+    <ProjectConfiguration Include="Release|x64">
+      <Configuration>Release</Configuration>
+      <Platform>x64</Platform>
+    </ProjectConfiguration>
+  </ItemGroup>
+  <PropertyGroup Label="Configuration">
+    <ConfigurationType>DynamicLibrary</ConfigurationType>
+    <PlatformToolset>v140</PlatformToolset>
```
Comment (asklar) [inline windows/DateTimePickerWindows/DateTimePickerWindows.vcxproj:49]: can we not? :)

File: windows/DateTimePickerWindows/readme.txt [hunk 7]
```diff
@@ -0,0 +1,23 @@
+========================================================================
+    C++/WinRT DateTimePickerWindows Project Overview
+========================================================================
+
+This project demonstrates how to get started authoring Windows Runtime
```
Comment (asklar) [inline windows/DateTimePickerWindows/readme.txt:5]: update placeholder or remove

File: package.json [hunk 8]
```diff
@@ -68,7 +69,8 @@
   },
   "peerDependencies": {
     "react": ">=16.11",
-    "react-native": ">=0.60"
+    "react-native": ">=0.60",
+    "react-native-windows": "^0.62.0-preview.1"
```
[the remark this answers is not in the document]
Comment (kaiguo) [reply package.json:73]: Can you also include react-native-windows in devDependencies, otherwise it won't restore rnw when doing yarn install.

File: example/windows/datePickerTest/ReactPackageProvider.cpp [hunk 9]
```diff
@@ -0,0 +1,20 @@
+#include "pch.h"
```
[the remark this answers is not in the document]
Comment (rectified95) [reply example/windows/datePickerTest/ReactPackageProvider.cpp:1]: I'm not sure. Looking at Kai's Picker PR (https://github.com/react-native-community/react-native-picker/pull/58/files#diff-04c6e90faac2675aa89e2176d2eec7d8) the license is not present everywhere. I can add it to the new files I'm creating.

[the remark this answers is not in the document]
Comment (asklar) [reply example/windows/datePickerTest/ReactPackageProvider.cpp:1]: I'm not sure either but agree we should be consistent across all community modules :) you might want to check what CELA guidance is or if @cglein has it.

[the remark this answers is not in the document]
Comment (kaiguo) [reply example/windows/datePickerTest/ReactPackageProvider.cpp:1]: Yeah I'm not sure either 😂
I think we should have it in all files, I'll update the template to include it.

File: package.json (hunk 8 above)
[the remark this answers is not in the document]
Comment (rectified95) [reply package.json:73]: Sure, will change. You meant 'more strict' though, right?

File: example/windows/datePickerTest/Package.appxmanifest [hunk 10]
```diff
@@ -0,0 +1,49 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+
+<Package
+  xmlns="http://schemas.microsoft.com/appx/manifest/foundation/windows10"
+  xmlns:mp="http://schemas.microsoft.com/appx/2014/phone/manifest"
+  xmlns:uap="http://schemas.microsoft.com/appx/manifest/uap/windows10"
+  IgnorableNamespaces="uap mp">
+
+ <Identity
+    Name="1a8165d9-9e81-4962-b325-9b4ca067b645"
```
[the remark this answers is not in the document]
Comment (rectified95) [reply example/windows/datePickerTest/Package.appxmanifest:10]: just change it to RNWDatePickerTest or is it possible to define it as a constant here?

[the remark this answers is not in the document]
Comment (asklar) [reply example/windows/datePickerTest/Package.appxmanifest:10]: this is just used by the package manager (and there's no logic for the build system to pull it out of anywhere else) so just change it to the string "RNWDatePickerTest" or something like that

File: react-native.config.js [hunk 11]
```diff
@@ -13,3 +13,13 @@ module.exports = {
     },
   },
 };
+
+const windowsSwitch = '--use-react-native-windows';
+
+if (process.argv.includes(windowsSwitch)) {
+  process.argv = process.argv.filter(arg => arg !== windowsSwitch);
```
[the remark this answers is not in the document]
Comment (rectified95) [reply react-native.config.js:20]: I don't understand - right now I substitute the 'windowsSwitch' in argv for the '--config' option - how should I use the filtered argv? (btw, this is code pasted from our migration tutorial)

File: example/windows/datePickerTest/ReactPackageProvider.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (asklar) [reply example/windows/datePickerTest/ReactPackageProvider.cpp:1]: not in the template since that's what apps get - but in any code that we write ourselves (RNW core, viewmanagers, native modules, and community modules), makes sense to me

File: react-native.config.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (asklar) [reply react-native.config.js:20]: So my comment was along the lines of: who is using process.argv and expecting to see the --config option instead of the --use-react-native-windows option? can they receive an arguments parameter and we pass them a copy of process.argv instead of modifying process.argv directly?

File: src/datetimepicker.windows.js [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+  rnDatePicker: {
+    height: 32,
+    width: 150,
+  },
+});
+
+const RNDateTimePickerWindows = requireNativeComponent('RNDateTimePickerWindows');
+
+export default class DateTimePickerWindows extends React.Component<WindowsNativeProps> {
+  static defaultProps: WindowsNativeProps = {
```
[the remark this answers is not in the document]
Comment (rectified95) [reply src/datetimepicker.windows.js:23]: "The CalendarDatePicker uses the correct calendar for your app's default language, or you can set the CalendarIdentifier property to use a specific calendar system" (https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.calendardatepicker?view=winrt-18362#globalization)

File: src/datetimepicker.windows.js [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+    dateFormat: 'day month year',
```
[the remark this answers is not in the document]
Comment (rectified95) [reply src/datetimepicker.windows.js:24]: The current implementation sets the default value to 'dayofweek day month'. I changed it during testing, can remove altogether.

File: src/datetimepicker.windows.js [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+      selectedDate: this.props.value
+        ? this.props.value.getTime()
+        : undefined, // time in milliseconds
+      style: [styles.rnDatePicker, this.props.style],
+    };
+
+    // The Date object returns timezone in minutes. Convert that to seconds
+    // and multiply by -1 so that the offset can be added to GMT time to get
+    // the correct value on the native side.
+    const timeZoneOffsetInSeconds = (this.props.timeZoneOffsetInSeconds != undefined) ?
```
[the remark this answers is not in the document]
Comment (rectified95) [reply src/datetimepicker.windows.js:49]: "(this.props.timeZoneOffsetInSeconds) ?" gave me a flow warning of 'sketchy null check for Number'. This prop is optional so I just wanted to check if it exists, while allowing 0.

File: src/datetimepicker.windows.js [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+      onChange: this.props.onChange,
+      placeholderText: this.props.placeholderText,
+      selectedDate: this.props.value
+        ? this.props.value.getTime()
+        : undefined, // time in milliseconds
+      style: [styles.rnDatePicker, this.props.style],
+    };
+
+    // The Date object returns timezone in minutes. Convert that to seconds
+    // and multiply by -1 so that the offset can be added to GMT time to get
```
Comment (asklar) [inline src/datetimepicker.windows.js:47]: actually you don't want GMT you want UTC+0 (since GMT can go into summer time too)

File: windows/DateTimePickerWindows/DateTimePickerWindows.vcxproj (hunk 6 above)
Comment (rectified95) [reply windows/DateTimePickerWindows/DateTimePickerWindows.vcxproj:49]: There's a condition declaring v142 for VS2019 in the line below this one. Shall I delete the lines for v140/141? (This is auto-generated btw)

File: src/datetimepicker.windows.js (hunk 15 above)
Comment (rectified95) [reply src/datetimepicker.windows.js:47]: Copied from current implementation. I think this comment uses GMT liberally and just means timezone-aware time - I don't see any references to UTC/GMT here. What do you think?

File: example/windows/datePickerTest/ReactPackageProvider.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (rectified95) [reply example/windows/datePickerTest/ReactPackageProvider.cpp:1]: @chrisglein should we put the license at the top of all new files?

File: windows/DateTimePickerWindows/DateTimePickerView.h (hunk 4 above)
Comment (rectified95) [reply windows/DateTimePickerWindows/DateTimePickerView.h:23]: Should I create aliases for all namespaces I reference in this project?

Comment (asklar) [reply windows/DateTimePickerWindows/DateTimePickerView.h:23]: @rectified95 you should only need to create aliases for the top level namespaces that will change with WinUI3. See this file: https://github.com/microsoft/react-native-windows/pull/4766/files#diff-fed223fe01e306d5faa2fb1af7633252R98

namespace xaml = winrt::Microsoft::UI::Xaml;
namespace comp = winrt::Microsoft::UI::Composition;
namespace winrt {
namespace system = winrt::Microsoft::System;
using ColorHelper = winrt::Microsoft::UI::ColorHelper;
using Colors = winrt::Microsoft::UI::Colors;
} // namespace winrt

File: windows/DateTimePickerWindows/DateTimePickerWindows.vcxproj (hunk 6 above)
Comment (asklar) [reply windows/DateTimePickerWindows/DateTimePickerWindows.vcxproj:49]: yeah and we should update the template to remove 140/141, that's probably something we missed when we migrated to 2019/142.

File: src/datetimepicker.windows.js (hunk 15 above)
Comment (asklar) [reply src/datetimepicker.windows.js:47]: UTC is the correct terminology :) let's update it wherever it applies. Having worked on timezones, it's important to be accurate :)

File: src/datetimepicker.windows.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (asklar) [reply src/datetimepicker.windows.js:23]: how is this exposed to RN apps though? what property do they set on the js side that will set the XAML CalendarIdentifier property for them?

File: src/datetimepicker.windows.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (asklar) [reply src/datetimepicker.windows.js:24]: Could you validate that the right behavior happens when you run the app in a locale with a different date format?
- different language (to test day/month names)
- different number system (e.g. Arabic numerals)
- different day/month/year order (e.g. Spanish uses ddMMyyyy instead of MMddyyyy)
- different calendar (e.g. Japanese Imperial calendar)

File: src/datetimepicker.windows.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (rectified95) [reply src/datetimepicker.windows.js:23]: You're right, we don't currently expose it; I migrated the supported set of properties from Core. 
Also, I meant that it will adjust automatically to the locale - is that not enough? I can see how someone might want to change eg. the date format, but countries only ever use one version of the calendar

File: example/windows/datePickerTest/pch.h [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+#define NOMINMAX
+
+#include <hstring.h>
+#include <restrictederrorinfo.h>
+#include <unknwn.h>
+#include <windows.h>
+#include <winrt/Windows.ApplicationModel.Activation.h>
+#include <winrt/Windows.Foundation.Collections.h>
+#include <winrt/Windows.Foundation.h>
+#include <winrt/Windows.UI.Xaml.Controls.Primitives.h>
```
[the remark this answers is not in the document]
Comment (rectified95) [reply example/windows/datePickerTest/pch.h:12]: Actually, the module itself doesn't import any XAML types; its pch.h:
#pragma once
#include <unknwn.h>
#include <winrt/Windows.Foundation.h>
#include <winrt/Windows.Foundation.Collections.h>

File: windows/DateTimePickerWindows/DateTimePickerView.h (hunk 4 above)
Comment (rectified95) [reply windows/DateTimePickerWindows/DateTimePickerView.h:23]: Btw, I'm using winrt::**Windows**::UI::Xaml, not winrt::Microsoft::UI::Xaml - the WinUI3 PR changes the latter, so will this code be affected by it?

Comment (asklar) [reply windows/DateTimePickerWindows/DateTimePickerView.h:23]: right - you can just define your alias to point xaml = winrt::Windows::UI::Xaml and then use xaml. Then when we are ready to make this module WinUI3 compatible, we just need to change what the alias points to (instead of having to change every place where we use a Xaml type)

File: src/datetimepicker.windows.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (rectified95) [reply src/datetimepicker.windows.js:24]: @asklar Just changed my region to Poland - can see different time/date formats and translated names **in explorer**, but not in the test app, even after reboot. 
However, the same happens when checking the DatePicker example in RNTester, so I'm not introducing a regression. Can this be a broader issues with our platform?

[the remark this answers is not in the document]
Comment (asklar) [reply src/datetimepicker.windows.js:24]: @rectified95 yes, it's possible this is a broader platform issue. Could you please file a bug in our repo with these details/screenshots? we should understand how RN apps expect to handle i18n and what role should RNW play in it. Normally you get these things for free in XAML so I'm surprised we don't see the right things happen :(
@chrisglein FYI

**Comment by rectified95** [comment]
@vonovak 
Looking at CHANGELOG.md, I see that not all commits are recorded there. 
I still guess that I should mention this change - shall I put it under 2.3.2, bump the version to 2.3.3, or are you responsible for versioning?

Also, are you the only one who can merge this PR?

File: src/datetimepicker.windows.js [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+    dateFormat: 'day month year',
+    dayOfWeekFormat: '{dayofweek.abbreviated(3)}'
+  };
+
+  constructor(props: WindowsNativeProps) {
+    super(props);
+  }
+  render() {
+    const props = {
+      dayOfWeekFormat: this.props.dayOfWeekFormat,
```
[the remark this answers is not in the document]
Comment (rectified95) [reply src/datetimepicker.windows.js:30]: @kaiguo I've documented all the new props, and even added a nice picture to the docs ;)
Also, added a section about how to link this module to a windows test app.

File: src/datetimepicker.windows.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (rectified95) [reply src/datetimepicker.windows.js]: opened issue: https://github.com/microsoft/react-native-windows/issues/4791

File: src/datetimepicker.windows.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (rectified95) [reply src/datetimepicker.windows.js]: opened issue: microsoft/react-native-windows#4791

File: example/windows/datePickerTest/ReactPackageProvider.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (rectified95) [reply example/windows/datePickerTest/ReactPackageProvider.cpp:1]: I've added the license to all .h/cpp/.idl files. WIll push with the final patch if there are more comments.

File: example/App.js [hunk 18]
```diff
@@ -12,15 +12,30 @@ import {
 import DateTimePicker from '@react-native-community/datetimepicker';
 import {Header, Colors} from 'react-native/Libraries/NewAppScreen';
 import React, {useState} from 'react';
+import {Picker} from 'react-native-windows';
 import moment from 'moment';
+import {DAY_OF_WEEK} from '../src/constants';
 
 const App = () => {
-  const [date, setDate] = useState(new Date(1598051730000));
```
Comment (vonovak) [inline example/App.js:18]: is removing the `1598051730000` part a problem for some reason? 

The end-to-end tests for ios and android, albeit currently not executed, rely on this. I'm planning to reintroduce them soon, so keeping this would make that a little easier. Thanks!

File: src/datetimepicker.windows.js [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+        {...props}
+        onChange={this._onChange}
+        timeZoneOffsetInSeconds={timeZoneOffsetInSeconds}
+      />
+    );
+  }
+
+  _onChange = (event: WindowsDatePickerChangeEvent) => {
+    this.props.onChange &&
+      this.props.onChange(event, new Date(+event.nativeEvent.newDate));
```
Comment (vonovak) [inline src/datetimepicker.windows.js:66]: just wondering, what is the `+` for? Thanks.

File: metro.config.windows.js [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+  resolver: {
+    extraNodeModules: {
+      // Redirect react-native to react-native-windows
+      'react-native': rnwPath,
+      'react-native-windows': rnwPath,
+    },
+    // Include the macos platform in addition to the defaults because the fork includes macos, but doesn't declare it
+    platforms: ['ios', 'android', 'windesktop', 'windows', 'web', 'macos'],
+    // Since there are multiple copies of react-native, we need to ensure that metro only sees one of them
+    // This should go in RN 0.61 when haste is removed
```
Comment (vonovak) [inline metro.config.windows.js:28]: // This should go in RN 0.61 when haste is removed

is this relevant now? Should it be removed / edited..?

**Comment by vonovak** [comment]
@rectified95 yes, the changelog is for now edited manually, I'd like to automate releases at some point soon but we're not there yet.

I think you don't need to worry about the changelog, I'll put it in there. I'm not the only one who can merge PRs but I have been the most active one here recently it seems. 

Regarding maintenance, do you have an example somewhere of how this functionality can be tested? Preferably end-to-end; thank you! Would the author or someone from the reviewers be open to becoming a maintainer? (for the windows part at least, but for I'd be very open to improvements outside of windows too) Thank you!

File: example/App.js (hunk 18 above)
Comment (rectified95) [reply example/App.js:18]: I removed it so that the 'text placeholder' prop in the windows version gets displayed initially, before the control is clicked and a date is selected.
image

I can restore it and add a 'windowsDate' prop to achieve the same thing, though.

File: src/datetimepicker.windows.js (hunk 19 above)
Comment (rectified95) [reply src/datetimepicker.windows.js:66]: tbh, I copied this bit over from the legacy implementation; I think it was about forcing a conversion from string to int, but I don't think it's necessary

File: metro.config.windows.js (hunk 20 above)
Comment (rectified95) [reply metro.config.windows.js:28]: I can change the number to 62, but we still need to override the metro config for out-of-tree platforms. There is an active PR to make it unnecessary
https://github.com/react-native-community/cli/pull/1115
(partial fix) https://github.com/microsoft/react-native-windows/issues/4081

**Comment by rectified95** [comment]
@vonovak  Thanks for your comments. 
About testing, you can run the test app on Windows as outlined in the PR description.

We'll get back to you about becoming maintainers; for now I can say that I'll fix bugs we spot after the merge.

File: README.md [hunk 21]
```diff
@@ -55,6 +63,22 @@ React Native date & time picker component for iOS and Android
 ## Requirements
 
 - Xcode >= 11; please open a PR if you need to support building with older Xcode versions - see [#133](https://github.com/react-native-community/datetimepicker/issues/133). Please note that we support iOS >= 10 but you need Xcode 11 to build.
+### Windows
```
Comment (kaiguo) [inline README.md:66]: Sorry I missed this part, can you move this down to the "Manual installation" section?

File: src/datetimepicker.windows.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (asklar) [reply src/datetimepicker.windows.js]: @rectified95 it's actually quite common in some countries to use two calendars, e.g. Japan with the Imperial  Calendar for government events and Gregorian for everyday affairs. Same with Israel with religious events with the Jewish Calendar and the Gregorian for secular events.

I'm fine with this going in without this support right now, but please open an issue to track adding a property to expose it what calendar to use. @chrisglein @harinikmsft FYI

[the remark this answers is not in the document]
Comment (kaiguo) [reply src/datetimepicker.windows.js]: I think they have `locale` for this? It's iOS only but we can probably make it work for Windows. https://github.com/react-native-community/datetimepicker#locale-optional-ios-only

[the remark this answers is not in the document]
Comment (rectified95) [reply src/datetimepicker.windows.js]: @asklar I've created microsoft/react-native-windows#4879

File: README.md [hunk 22]
```diff
@@ -37,6 +41,10 @@ React Native date & time picker component for iOS and Android
     - [`maximumDate` (`optional`)](#maximumdate-optional)
     - [`minimumDate` (`optional`)](#minimumdate-optional)
     - [`timeZoneOffsetInMinutes` (`optional`, `iOS only`)](#timezoneoffsetinminutes-optional-ios-only)
+    - [`timeZoneOffsetInSeconds` (`optional`, `Windows only`)](#timezoneoffsetinsecond-optional-windows-only)
```
Comment (asklar) [inline README.md:44]: curious why have a separate property that is Windows only for this if we can piggyback on the iOS property?

File: README.md [hunk 23]
```diff
@@ -37,6 +41,10 @@ React Native date & time picker component for iOS and Android
...
+    - [`dayOfWeekFormat` (`optional`, `Windows only`)](#dayOfWeekFormat-optional-windows-only)
+    - [`dateFormat` (`optional`, `Windows only`)](#dateFormat-optional-windows-only)
+    - [`firstDayOfWeek` (`optional`, `Windows only`)](#firstDayOfWeek-optional-windows-only)
```
Comment (asklar) [inline README.md:47]: these 3 properties are part of locale data, do we really need separate properties that are Windows only?

File: README.md [hunk 24]
```diff
@@ -234,6 +258,45 @@ Allows changing of the timeZone of the date picker. By default it uses the devic
 <RNDateTimePicker timeZoneOffsetInMinutes={60} />
 ```
 
+#### `timeZoneOffsetInSeconds` (`optional`, `Windows only`)
+
+Allows changing of the time zone of the date picker. By default it uses the device's time zone.
+
+```js
+// GMT+1
```
Comment (asklar) [inline README.md:266]: UTC

File: windows/DateTimePickerWindows/DateTimePickerWindows.vcxproj [hunk 25]
```diff
@@ -0,0 +1,169 @@
...
+    <ProjectGuid>{0986a4db-8e72-4bb7-ae32-7d9df1758a9d}</ProjectGuid>
+    <ProjectName>DateTimePickerWindows</ProjectName>
+    <RootNamespace>DateTimePicker</RootNamespace>
+    <DefaultLanguage>en-US</DefaultLanguage>
+    <MinimumVisualStudioVersion>14.0</MinimumVisualStudioVersion>
+    <AppContainerApplication>true</AppContainerApplication>
+    <ApplicationType>Windows Store</ApplicationType>
+    <ApplicationTypeRevision>10.0</ApplicationTypeRevision>
+    <WindowsTargetPlatformVersion Condition=" '$(WindowsTargetPlatformVersion)' == '' ">10.0.18362.0</WindowsTargetPlatformVersion>
+    <WindowsTargetPlatformMinVersion>10.0.17134.0</WindowsTargetPlatformMinVersion>
```
Comment (asklar) [inline windows/DateTimePickerWindows/DateTimePickerWindows.vcxproj:18]: Don't we have to support 16299?

**Comment by vonovak** [comment]
hello guys, please ping me when this is done, just so I know explicitly that the review is done on your side, thank you!

File: example/App.js [hunk 26]
```diff
@@ -53,93 +68,207 @@ const App = () => {
...
-                />
-              </View>
-              <View style={styles.button}>
-                <Button
-                  testID="datePickerButtonSpinner"
-                  onPress={showDatepickerSpinner}
-                  title="Show date picker spinner!"
-                />
+  if (Platform.OS !== 'windows') {
+    return (
```
Comment (NickGerleman) [inline example/App.js:72]: Can we share more between platforms?

File: react-native.config.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (NickGerleman) [reply react-native.config.js:20]: Is there a pattern for solving the Metro issue used in other community modules? Was under the impression we could share metro configs (though I think things get wonky if you add macos with windows as well).

File: src/datetimepicker.windows.js [hunk 27]
```diff
@@ -0,0 +1,68 @@
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
```
Comment (NickGerleman) [inline src/datetimepicker.windows.js:3]: This came up in another repo. Not sure it's kosher to put msft copyright in a community repo we don't matinain.

File: README.md (hunk 23 above)
Comment (asklar) [reply README.md:47]: Filed #161

File: README.md (hunk 22 above)
Comment (asklar) [reply README.md:44]: Filed #161

File: src/datetimepicker.windows.js [hunk 28]
```diff
@@ -0,0 +1,68 @@
...
+    height: 32,
+    width: 150,
+  },
+});
+
+const RNDateTimePickerWindows = requireNativeComponent(
+  'RNDateTimePickerWindows',
+);
+
+export default class DateTimePickerWindows extends React.Component<WindowsNativeProps> {
```
Comment (NickGerleman) [inline src/datetimepicker.windows.js:24]: `WindowsNativeProps` causes a bit of confusion for me. I've previously seen "native component properties" refer to the props being passed to `RNDateTimePickerWindows` instead of `DateTimePickerWindows`, since the former is considered the native component.

Future versions of Flow + TS types parameterize `requireNativeComponent` to take props as well which would likely take a `NativeProps` type.

File: src/datetimepicker.windows.js [hunk 29]
```diff
@@ -0,0 +1,68 @@
...
+      selectedDate: this.props.value ? this.props.value.getTime() : undefined, // time in milliseconds
+      style: [styles.rnDatePicker, this.props.style],
+    };
+
+    // The Date object returns timezone in minutes. Convert that to seconds
+    // and multiply by -1 so that the offset can be added to UTC+0 time to get
+    // the correct value on the native side.
+    const timeZoneOffsetInSeconds =
+      this.props.timeZoneOffsetInSeconds != undefined
+        ? this.props.timeZoneOffsetInSeconds
```
Comment (NickGerleman) [inline src/datetimepicker.windows.js:50]: Could we avoid the nested ternaries? This is a little bit hard to parse out.

File: package.json [hunk 30]
```diff
@@ -11,6 +11,7 @@
     "start": "react-native start",
     "start:android": "react-native run-android --no-jetifier",
     "start:ios": "react-native run-ios",
+    "start:windows": "react-native start --use-react-native-windows",
```
Comment (NickGerleman) [inline package.json:14]: Mentioned somthing in another comment, but can we share the metro config and use `run-windows`?

File: README.md (hunk 22 above)
Comment (rectified95) [reply README.md:44]: Discussed separately, resolving.

File: src/datetimepicker.windows.js (hunk 27 above)
Comment (rectified95) [reply src/datetimepicker.windows.js:3]: Alex/Kai told me in an earlier comment to add it to all these files. Can you link the discussion you're referencing?

Comment (NickGerleman) [reply src/datetimepicker.windows.js:3]: They'll likely know better on this. IIRC there was one community repo that asked that copyright headers be removed. @kaiguo do you remember the story there?

File: example/App.js (hunk 26 above)
Comment (rectified95) [reply example/App.js:72]: The example app demoes a few modes that we don't support - time picker (not yet) and spinner. Thought it'd be easier to just add 1 branch vs exclude all those tags for windows.

File: package.json (hunk 30 above)
Comment (rectified95) [reply package.json:14]: Not sure, but I think we still need to swap the metro config

Comment (NickGerleman) [reply package.json:14]: @acoates-ms did some work to allow a single metro config allowing multiple out of tree platforms for 0.62. Would you mind checking with him? Want to make sure we set the right patterns for future community modules based on 0.62.

File: windows/DateTimePickerWindows/DateTimePickerWindows.vcxproj (hunk 25 above)
Comment (rectified95) [reply windows/DateTimePickerWindows/DateTimePickerWindows.vcxproj]: The CalendarDatePicker class is supported in 16299, so I've changed it accordingly :)

File: src/datetimepicker.windows.js (hunk 27 above)
Comment (kaiguo) [reply src/datetimepicker.windows.js:3]: Hmm, I guess that's a @chrisglein question, should we put Microsoft copyright info in a community repo we don't own?

**Comment by rectified95** [comment]
@vonovak My colleagues have signed off on the PR.

I've also tried to set up a build workflow, but it seems like I don't have the privileges, so I've reverted it (that's what the text says if you click the 'Actions' tab in this repo). We can coordinate this as a separate PR.
Also, I want to reiterate that I'll fix bugs related to this check-in. :)

File: src/datetimepicker.windows.js (hunk 27 above)
Comment (chrisglein) [reply src/datetimepicker.windows.js:3]: Where we landed for the first couple repos (camera, video, etc.) is to leave it off. We match the pattern of the target repo. I'll check offline, but for now assume that's the case.

Comment (rectified95) [reply src/datetimepicker.windows.js:3]: Sounds good, the Android/iOS files in this repo have the MIT license from Facebook, so I'll keep ours for now.

**Comment by vonovak** [comment]
Thanks so much for this!

> I've also tried to set up a build workflow, but it seems like I don't have the privileges, so I've reverted it (that's what the text says if you click the 'Actions' tab in this repo)

that'd be nice to have. But for this repo, we're not using github actions but circleCI. Since circle runs for PRs too, I'm guessing you should be able to make changes to the config and see the results in a PR.

> Also, I want to reiterate that I'll fix bugs related to this check-in. :)

great, let me know if I can help

**Comment by rectified95** [comment]
@vonovak Thanks for being responsive and merging this! :)

There have been issues with certificates for release builds on Windows VMs in CircleCI. A colleague ran into this when adding a CI loop for the WebView module, and ended up having to use Github actions - for WebView they run alongside the CircleCI steps; is there a reason we cannot add GitHub actions for this project?
For reference: https://github.com/react-native-community/react-native-webview/pull/1358

**Comment by vonovak** [comment]
hi @rectified95 

> There have been issues with certificates for release builds on Windows VMs in CircleCI.

okay, I had no idea about that

> is there a reason we cannot add GitHub actions for this project?

I wanted to prefer using circle because we already use it so there is no entry barrier, but if there is a blocker there, I'm def not opposed to using actions too. I can give whoever wants to do this the necessary access rights.

**Comment by rectified95** [comment]
@vonovak You can give me the permissions, I'll try setting it up.
